[Dynamo] Replace `TorchRuntimeError` with `Unsupported`

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119599
* #119859
* #119838
* #119836
* #119835
* #119834

Fixes #118794

In order for eager mode and compiled mode to match in behavior, graph break instead of raising TorchRuntimeError.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

oulgen committed:
[Dynamo] Fix FakeTensor exception being hard error

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Fix FakeTensor exception being hard error"


Fixes #118794



[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Fix FakeTensor exception being hard error"


Fixes #118794



[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Fix FakeTensor exception being hard error"


Fixes #118794



[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Fix FakeTensor exception being hard error"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Fix FakeTensor exception being hard error"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1424,6 +1424,8 @@ def maybe_run_unsafe_fallback(error=None):
                 r = func(*args, **kwargs)
         except NotImplementedError as not_implemented_error:
             return maybe_run_unsafe_fallback(not_implemented_error)
+        except RuntimeError as e:
+            raise UnsupportedOperatorException(str(e)) from e
```
Do you have a smaller repro that demonstrates the error?

I don't think we want to do this: this is converting all RuntimeErrors into graph breaks. If there is a RuntimeError because someone implemented the FakeTensor kernel incorrectly, we want to know about it so that we can fix it.

oulgen committed:
Update on "[Dynamo] Fix FakeTensor exception being hard error"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Promote RuntimeError to Unsupported when the CALL* opcode is in a try-catch"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aorenste commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -581,6 +581,33 @@ def handle_graph_break(
+        except Unsupported:
+            raise
+        except RuntimeError as e:
+            if inst.offset is None or not hasattr(self.f_code, "co_exceptiontable"):
+                raise
+
+            from torch._dynamo.bytecode_transformation import parse_exception_table
+
+            table = parse_exception_table(self.f_code.co_exceptiontable)
+            for entry in table:
+                if entry.start <= inst.offset < entry.end:
+                    # If we are in a try block of a try-catch, promote the
+                    # exception to graph break
+                    raise Unsupported(str(e)) from e
+            raise
```
Consider pulling this code into a helper function. I can imagine other code wanting to know if a given instr is within an exception handling block.

oulgen replied:
Based on a simple code search, this is the only place we currently do this. Do you think it is valueable create the helper function now or if anything else needs it to do it at that time?

aorenste replied:
IMO yes - it encapsulates the functionality and also provides immediate documentation on intent for code readers (`self.is_instr_in_exception_handler(instr)` is a lot more readable what the code is doing...). But up to you - it wouldn't hold me up from accepting.

oulgen replied:
Done

aorenste reviewed (commented):
Also consider adding a smaller unit test. Do we have intrinsics that allow us to build tests with specific conditions like "pretend C code threw an exception", "force a graph break here", etc?

oulgen commented:
> Also consider adding a smaller unit test. Do we have intrinsics that allow us to build tests with specific conditions like "pretend C code threw an exception", "force a graph break here", etc?

Nothing I am aware of, @zou3519 @anijain2305 ?

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -581,6 +581,33 @@ def handle_graph_break(
+
+            table = parse_exception_table(self.f_code.co_exceptiontable)
+            for entry in table:
+                if entry.start <= inst.offset < entry.end:
+                    # If we are in a try block of a try-catch, promote the
+                    # exception to graph break
+                    raise Unsupported(str(e)) from e
```
I want a sanity check from the Dynamo folks (@anijain2305, @yanboliang) on this. If an exception gets thrown inside of a try-catch block, we end up promoting it to a graph break (instead of allowing it through). I hope this is not another "suppress_errors=True" situation; the argument is that (1) most people won't throw their code under try-catch and (2) we should only try to convert errors that are not internal dynamo errors into graph breaks

zou3519 replied:
If we are worried about a suppress_errors=True situation, then maybe we could limit this change to just errors being thrown from FakeTensor prop?

oulgen committed:
Update on "[Dynamo] Promote RuntimeError to Unsupported when the CALL* opcode is in a try-catch"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
One bummer I discovered while looking at the failing tests is that co_exceptiontable is added in python 3.11. It will be hard to replicate this with 3.8's functionality. @zou3519 do we have a stance on our reliability goals for python versioning?

zou3519 commented:
@oulgen pytorch supports python 3.8-3.11 (and soon 3.12). We generally make a decision based on when the EOL for the python version is and its popularity. Though, is this a pre-3.11 problem? Pre-3.11, there is a SETUP_FINALLY bytecode at the beginning of the try-catch block, which should give us similar information.

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -581,6 +581,26 @@ def handle_graph_break(
+            return inner_fn(self, inst)
+        except Unsupported:
+            raise
+        except RuntimeError as e:
```
Without this PR, Dynamo will convert the RuntimeError into either a InternalTorchDynamoError or TorchRuntimeError. We should only convert RuntimeErrors that would have been converted into TorchRuntimeError into graph breaks; InternalTorchDynamoError RuntimeErrors are Dynamo bugs and should not be silently ignored.

(that being said I am not sure where this logic is)

jansel reviewed (changes requested):
I worry thing will be hiding bugs inside dynamo.  We shouldn't be silently ignoring errors.

Can we do a more narrow fix for the issue?

oulgen committed:
Update on "[Dynamo] Promote RuntimeError to Unsupported when the CALL* opcode is in a try-catch"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
@jansel I updated the PR to only catch `TorchRuntimeError ` and not catch `InternalTorchDynamoError `. Do you think this would be enough along with only doing this if the code is in a try part of try-catch?

The problem I am trying to solve is user code needs to be able to catch exceptions being thrown torch.xyz ops. However, currently dynamo ends up throwing `TorchRuntimeError` so we never re-enter user code.

jansel commented:
Graph break on TorchRuntimeError seems fine, since that means eager mode will throw an exception.  Perhaps we should just replace TorchRuntimeError with `unimplemented()` though so eager/compile have the same exception behavior.

oulgen commented:
This PR effectively does that except only when the `TorchRuntimeError` is thrown in exception region. Would you prefer me to kill all `TorchRuntimeError` replace with unimplemented? I wonder whether that might have other repercussions

jansel commented:
Yes, this whole catch and change the exception thing is confusing, and you might have missed some places.

oulgen committed:
Update on "[Dynamo] Promote RuntimeError to Unsupported when the CALL* opcode is in a try-catch"


Fixes #118794



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented on test/dynamo/test_logging.py:
```diff
@@ -132,26 +132,21 @@ def test_dynamo_debug_default_off_artifacts(self, records):
         self.assertEqual(len([r for r in records if ".__output_code" in r.name]), 0)
 
-    @make_logging_test()
+    @make_logging_test(graph_breaks=True)
```
not sure whether there's a better way to test this now that we can't really error from user code

oulgen committed:
Update on "[Dynamo] Replace `TorchRuntimeError` with `Unsupported`"


Fixes #118794

In order for eager mode and compiled mode to match in behavior, graph break instead of raising TorchRuntimeError.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel reviewed (changes requested):
Actually maybe it would be clearer to make TorchRuntimeError a subclass of Unsupported, so it still results in a graph break, but we have better error messages about why we are graph breaking.

Looks like many tests are failing.

oulgen committed:
Update on "[Dynamo] Replace `TorchRuntimeError` with `Unsupported`"


Fixes #118794

In order for eager mode and compiled mode to match in behavior, graph break instead of raising TorchRuntimeError.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
I'm inclined to abandon this convert to graph break solution and find a proper solution. Concisely, the problem is

```
def foo(x):
    try:
         return torch.foo(x) # throws runtime error
    except RuntimeError:
         return x
```

needs to work in dynamo but currently FakeTensor prop throws a RuntimeException that user code cannot catch as it happens during tracing.

jansel commented:
IMO TorchRuntimeError should *always* graph break (by being a subclass of graph break).  My only concern was getting better error messages in explain/fullgraph=True.
